keps/sig-network/20191020-default-deny-pod-traffic

```diff
@@ -0,0 +1,270 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Make pod-to-pod communication in Kubernetes secure by default.
+Pods should require an explicit NetworkPolicy ingress configuration,
+in order to receive traffic.
+
+This proposal explicitly targets ingress traffic.
+Egress traffic is not _currently_ being targeted,
+not is not yet deemed out of scope.
+
+## Motivation
+
+Kubernetes and Kubernetes workloads are notoriously easy to leave insecure.
+Many users are not aware of relevant settings (like NetworkPolicy),
+or choose not to configure them due to friction and learning curves.
+This is a systemic pattern,
+NetworkPolicy is one example.
+
+TODO pull examples
+
+### Goals
+
+* Restrict pod ingress by default on a cluster.
+
+### Non-Goals
+
+What is out of scope for this KEP?
+Listing non-goals helps to focus discussion and make progress.
+
+## Proposal
+
+There are multiple options for design/implementation.
+
+Restricting network access requires a CNI driver that supports NetworkPolict.
+All options will require codifying NetworkPolicy support as a requirement,
+or creating a clear user experience when using a CNI driver that does not support NetworkPolicy.
+
+### Option 1: Change default NetworkPolicy behavior
+
+The most conceptually simple option is to change default NetworkPolicy behavior.
+Currently,
+NetworkPolicy allows all traffic to pods that are not selected by a NetworkPolicy. [1]
+
+If all pods are treated as isolated,
+users must explicitly create NetworkPolicy objects
+(either as needed, or by allowing all ingress and _explicitly_ enabling the existing behavior).
+This would potentially require a v2 API, due to changes in behavior (if not the API itself).
+
+### Option 2: Inject explicit deny-all NetworkPolicy objects
+
+If making changes to NetworkPolicy API is too undesirable,
+we could automatically apply NetworkPolicy objects to achieve the same effect.
+
+There is no precedent(?) or mechanism(?) for performing pre-creation actions
+(IE the opposite of finalizers).
```

Comment:
OpenShift has the "[default project template](https://docs.openshift.com/container-platform/4.2/applications/projects/configuring-project-creation.html#modifying-template-for-new-projects_configuring-project-creation)", although it only applies to user-provisioned namespaces, not administrative namespaces.

The nice thing about this sort of approach is that (given the right design) the administrator could use this for more than just "default-deny". Eg, they could automatically add an "allow-from-monitoring-namespace" policy to every namespace, etc. And they could choose to add default egress policies as well.

Reply:
> E.g. user can edit any NetworkPolicy except the "default" instance?

oh, I wasn't suggesting that the "allow from monitoring namespace" would be not removable by the project admin. It was more that, if you know every namespace is going to *want* to have that policy, then the admin could cause it to be auto-created for them.